### ISSUE TITLE
Roadmap for interpretability of fairness metrics?

### ISSUE THREAD

**Issue by hildeweerts** [fairlearn/fairlearn#458]
The correct interpretation of almost any fairness metric relies on which label is positive and which sensitive groups are compared. At the moment, for most metrics, these considerations are implicit in the API. For example, all confusion_matrix related performance metrics will return labels as: “those that appear at least once in y_true or y_pred are used in sorted order” due to the dependency on scikit-learn, but this is not explicitly mentioned in the fairlearn documentation. 

I was wondering whether there are plans to help users (particularly those who do not use the dashboard) interpret the metrics correctly.

I do not have a strong opinion on what it should look like, but I can imagine e.g. consistent usage of ``pos_label`` and an equivalent argument for privileged group membership might help (although the latter is tricky for polytomous sensitive features). Or perhaps some accompanying textual information, e.g. if demographic_parity_difference = 0.10, one could retrieve an 'explanation' along the lines of "selection rate of group=1 (largest) is 0.10 higher than the selection rate of group=2 (smallest)".

**Comment by riedgar-ms** [comment]
Thank you for the feedback. The matter of `pos_label` has certainly come up recently... I'm not sure but I think the reworked metrics API will actually pass that through to the underlying method. Having said that, the fact that you're asking about it means that (even if it would be passed through) it's not documented properly.

The point about a privileged group is an interesting one. Our metrics are always based on max and min, rather than a particular group. But different groups could get the max and min by different metrics. We probably need to support both - i.e. if `privileged_group=None` then do what we do now, and base on max and min. Otherwise, compare to the specified group (and concoct some way of handling the case where the privileged group might not be the max or min - easy for differences, ratios are not so obvious).

**Comment by hildeweerts** [comment]
On a related note, disparities between more than two sensitive groups are very difficult to capture in a single number. To be honest, I cannot really think of a practical scenario in which the min/max difference would be sufficient for me to judge the fairness of a model, let alone compare several models. You would always like to know between *which* groups the disparity is large, what the performance is like for the other groups that are not min/max, etc.

**Comment by romanlutz** [comment]
Thanks for raising the issue @hildeweerts !

There's indeed a lot to be improved in terms of the documentation. Ideally our documentation (API reference) should be clear enough that people understand what they're actually calling. Providing a string "explanation" is a pattern I haven't seen before, but if it's something people find useful we should consider it. As a user I usually don't use things unless I have an idea of what it does, which is why from my perspective the documentation improvements sound like the obvious place to fix this.

I agree that it gets difficult with the metrics when comparing more than two groups. That's pretty much the value proposition of the dashboard single model view. Beyond that, I couldn't agree more that a single metric (e.g. min/max difference) isn't sufficient. I hope the documentation doesn't say that anywhere. We should always take a look at the entire picture, but that's hard with just numbers and that's where the visualizations come in.

The group summaries do provide you with everything you need to make any sort of comparisons, though, in case you really want to do this without visualizations
```python
>>> from fairlearn.metrics import accuracy_score_group_summary
>>> y_true = [1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0]
>>> y_pred = [1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0]
>>> sensitive_features = ["a", "b", "b", "a", "a", "a", "b", "b", "a", "b", "b", "b", "a", "b"]
>>> accuracy_score_group_summary(y_true, y_pred, sensitive_features=sensitive_features)
{'overall': 0.7142857142857143, 'by_group': {'a': 0.8333333333333334, 'b': 0.625}}
```

Even if there are more groups you could always figure out specific differences between certain groups that are relevant to your context.

In terms of follow-ups I think it makes sense to iterate on the documentation and user guide. There's currently a PR out #451 that adds content to the user guide for metrics. That may help as a first step, but I don't see this as the final piece. @hildeweerts thanks again for raising this. It's really helpful to hear what is hard to understand, what you would have expected to see, etc. 

If you have a concrete suggestion please also feel free to submit a PR. We're happy to take any contributions!

**Comment by riedgar-ms** [comment]
As I think a bit more, we would probably want to have users select the group via a `selected_group` argument, rather than `privileged_group` since they may be wanting to view comparisons with a group which is getting particular bad results, rather than the best ones.

Other than that, I think that for the metrics code, our first order of business should be to:

- Ensure that `pos_value` is supported by all the classification metrics
- Add the `selected_group` argument for all metrics, and decide on a consistent way of handling differences and ratios in that case

Would that cover your main concerns there @hildeweerts ?

**Comment by hildeweerts** [comment]
> Ideally our documentation (API reference) should be clear enough that people understand what they're actually calling. Providing a string "explanation" is a pattern I haven't seen before, but if it's something people find useful we should consider it. As a user I usually don't use things unless I have an idea of what it does, which is why from my perspective the documentation improvements sound like the obvious place to fix this.

Yes, improving documentation would help. Nevertheless, even if you do know what you're calling it is very easy to make mistakes, especially in an exploratory phase, e.g. because you accidentally switched labels or something like that. I don't think textual explanations are necessarily the way to go, as @riedgar-ms mentioned there are many potential edge cases, but I do think it is something we need to think about.

> I agree that it gets difficult with the metrics when comparing more than two groups. That's pretty much the value proposition of the dashboard single model view. 

Using the dashboard is a nice way to circumvent these issues, but there are also many use cases in which the dashboard may not be sufficient or desirable. I wonder whether returning a single number should be the default behavior at all. I can also imagine returning the summary as default behavior while adding specific function arguments for aggregations. At least in this scenario, users would be forced to consciously think about what they are optimizing for.

> Ensure that pos_value is supported by all the classification metrics

Yes! It might already be supported for most if not all - so this would probably mostly be a documentation issue.

**Comment by MiroDudik** [comment]
These are my suggestions for next steps:
* improve documentation on when to use which metrics; we in fact have a more basic concern, which is to even expose the metrics that are already available (because they are hidden inside the metrics dictionaries right now)
* add `pos_label` to our own base metrics and make sure that it works with the metrics that we derive from `sklearn.metrics`
* in documentation, we should point out strongly that the aggregate metrics are just the initial diagnostic step and if you see that their value is large, you'll want to dig into the distribution of the base metric across the groups and so you should use `<metric>_group_summary`.
  * This was the rationale behind defining the worst-case versions of difference & ratio--they are meant to quantify the overall disparity rather than a disadvantage of a specific group. Thinking about disadvantage requires more nuance, so I'd much rather show the numbers across all groups and let that serve as a starting point for further exploration. 

I'm right now a bit less sure about the benefit of creating an additional functionality around `selected_group` or `privileged_group` as opposed to just using `<metric>_group_summary`. But this might be because I'm not quite sure about the specific metrics that you had in mind and the specific use case scenarios. What are those? (We could also cover that in a call if that's easier.)

**Comment by riedgar-ms** [comment]
For `pos_label`, I've just been experimenting, and it doesn't seem to work. I'll see if I can figure out how to fix that.

**Comment by romanlutz** [comment]
Related conversation going on at https://github.com/fairlearn/fairlearn-proposals/issues/12

**Comment by hildeweerts** [comment]
Close in favor of smaller issue #721. 

@MiroDudik I'm assuming your concerns re. `<metric>_group_summary` are not really relevant anymore with the new metrics API. If this is not the case please open a new issue to address it.
